docs/multi-agent-applications.md

```diff
@@ -0,0 +1,333 @@
+1. Single agent workflows — what most of this documentation covers
+2. [Agent delegation](#agent-delegation) — agents using another agent via tools
+3. [Programmatic agent hand-off](#programmatic-agent-hand-off) — one agent runs, then application code calls another agent
+4. [Graph based control flow](#pydanticai-graphs) — for the most complex cases, graph and a state machine can be used to control the execution of multiple agents
+
+Of course, you can combine multiple strategies in a single application.
+
+## Agent Delegation
+
+The agent delegates work to another agent, but then takes back control when that agent finishes.
+
+Since agents are stateless and designed to be global, you do not need to include the agent itself in agent [dependencies](dependencies.md).
+
+When doing so, you'll generally want to pass [`ctx.usage`][pydantic_ai.RunContext.usage] to the [`usage`][pydantic_ai.Agent.run] keyword argument of delegate agent (the agent called from within a tool) run so usage within that run counts towards the total usage of a parent agent run.
+
+!!! Multiple models
+    Agent delegation doesn't need to use the same model for each agent. If you choose to use different models within a run, calculating the monetary cost from the final [`result.usage()`][pydantic_ai.result.RunResult.usage] of the run will not be possible, but you can still use [`UsageLimits`][pydantic_ai.usage.UsageLimits] to avoid unexpected costs.
+
+```python {title="agent_delegation_simple.py"}
+from pydantic_ai import Agent, RunContext
+from pydantic_ai.usage import UsageLimits
+
+joke_agent = Agent(
+    'openai:gpt-4o',
+    system_prompt=(
+        'Use the "joke_factory" to generate some jokes, then choose the best. '
+        'You must return just a single joke.'
+    ),
+)
+
+
+@joke_agent.tool
+async def joke_factory(ctx: RunContext[None], count: int) -> list[str]:
+    r = await delegate_agent.run(f'Please generate {count} jokes.', usage=ctx.usage)
+    return r.data
+
+
+delegate_agent = Agent('gemini-1.5-flash', result_type=list[str])
+
+result = joke_agent.run_sync(
+    'Tell me a joke.', usage_limits=UsageLimits(request_limit=5, total_tokens_limit=300)
+)
+print(result.data)
+#> Did you hear about the toothpaste scandal? They called it Colgate.
+print(result.usage())
+"""
+Usage(
+    requests=3, request_tokens=204, response_tokens=24, total_tokens=228, details=None
+)
+"""
+```
+
+_(This example is complete, it can be run "as is")_
+
+The control flow for this example is pretty simple and can be summarised as follows:
+
+```mermaid
+graph TD
+  START --> joke_agent
+  joke_agent --> joke_factory["joke_factory (tool)"]
+  joke_factory --> delegate_agent
+  delegate_agent --> joke_factory
+  joke_factory --> joke_agent
+  joke_agent --> END
+```
+
+### Agent Delegation and dependencies.
+
+The delegate agent needs to either have the same [dependencies](dependencies.md) as the calling agent, or dependencies which are a subset of the calling agent's dependencies.
```

Comment:
This isn't strictly speaking true, it just needs to be the case that the required dependencies can be instantiated at the call site of the delegate agent. In particular, the delegate agent can make use of global dependencies, or functions that can be called in any context.

For example, if you have a globally-accessible `pydantic_settings` settings object that reads from environment, you could use that to get a connection string that you then use to instantiate a database connection, which is provided to the delegate agent via its `deps`, even though it didn't come from the parent agent's `deps`. And this feels like a somewhat realistic scenario to me; of course users _could_ pass a connection pool as a dep to the parent agent, but it's not necessary and would probably be skipped at least during prototyping.

Ultimately this isn't a big deal, but my point is just that I think if we don't make this point in some way, it _may_ leave users confused about the mental model they should have about how they can build dependencies. In particular, the deps from the parent agent _are not_ passed directly to the delegate agent through some complex and opaque mechanism (which might be my intuition from this sentence) — they are just instantiated in the call to the delegate agent.

Reply:
This reasoning makes sense to me, I am trying to use the multi-agent functionality and I am debating between passing all deps through the chain (from main to delegate) or only use the deps relevant to that agent.

Reply:
I've added a note about this.